### PR TITLE
CMake toolchain

### PR DESCRIPTION
**Pull request by ajneu**
Hi Alex,

been pondering this toolchain topic a bit. 

And I believe this new variant (see changed files) is superior to the previous version.

Nice things are for example... if the following 2 lines

``` cmake
set (CMAKE_SYSTEM_NAME "Generic")
cmake_force_c_compiler(  "${CROSS_COMPILE}gcc" GNU)
```

are called **before** the line with `project` (which is the case in this pull-req), then the binutils (ar, ranlib and even objcopy, etc.) are automatically found.
This is the change in `src/CMakeLists.txt` which now uses the automatically set `${CMAKE_OBJCOPY}`

Regards,
ajneu

### CODE REVIEW

File: cmake/toolchain_handling.cmake [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+else()
+  
+  set (CMAKE_SYSTEM_NAME "Generic")
+  
+  if ("${CROSS_COMPILE}" STREQUAL "")
+    set (CROSS_COMPILE "arm-none-eabi-")
+  endif()
+
+  include(CMakeForceCompiler)
+  cmake_force_c_compiler(  "${CROSS_COMPILE}gcc" GNU)
```
Comment (0xc0170) [inline cmake/toolchain_handling.cmake]: what version of cmake are you using ? Upcoming 3.5.0 deprecates this.

Comment (ajneu) [inline cmake/toolchain_handling.cmake]: Took me some time to understand what you mean here.
I've run it with cmake version 3.5.0-rc3 without any problems and without any warning.

Then I found: the deprecation is only mentioned here:
https://cmake.org/cmake/help/v3.5/module/CMakeForceCompiler.html?highlight=deprecated

But:
It's **not** deprecated in 3.5.x yet!!

So maby, for the future... [this variant](https://github.com/arobenko/embxx_on_rpi/pull/5) was good after all. But for that variant, one should edit `src/CMakeLists.txt` to be similar to the change in this pull-req here.

Or: take this pull-request here, and "fix" (or rather "beautify", by removing deprecation) once we find out what the CMake team has decided to do. (Possibly using a cmake-version conditional...)

**Comment by ajneu** [comment]
@0xc0170 By the way: Are you having any problems when using code from this pull-request?

**Comment by arobenko** [comment]
@ajneu 
I'm a bit reluctant to merge your pull request as it is now. I dislike the idea of having extra "cmake/toolchain_handling.cmake" file. It's functionality may be defined in place, without any include statements. Is there any need to print all the utilities being used? I think CMake itself prints what C and C++ compiler being used. It's also not clear whether `enable_language (ASM)` needs to be replaced with `enable_language (ASM-ATT)`. You just added a comment there without changing the statement.

**Comment by ajneu** [comment]
Thanks for your feedback.

> I'm a bit reluctant to merge your pull request as it is now. I dislike the idea of having extra "cmake/toolchain_handling.cmake" file. 

Yes, that was extracted, to make that main `CMakeLists.txt` smaller

> It's functionality may be defined in place, without any include statements. Is there any need to print all the utilities being used?

The printing is nice for debugging purposes. We see exactly which `ra`, `ranlib`, `objcopy`, `objdump` are being used. If one wants that printing, I'd keep `cmake/toolchain_handling.cmake`. If you don't want the diagnostic printing, then moving it all into the single `CMakeLists.txt`, it would look like this:

``` cmake
 cmake_minimum_required (VERSION 2.8)


if (CMAKE_TOOLCHAIN_FILE AND EXISTS ${CMAKE_TOOLCHAIN_FILE})
  message("==> Loading \${CMAKE_TOOLCHAIN_FILE} == ${CMAKE_TOOLCHAIN_FILE}")

else()

  set (CMAKE_SYSTEM_NAME "Generic")

  if ("${CROSS_COMPILE}" STREQUAL "")
    set (CROSS_COMPILE "arm-none-eabi-")
  endif()

  include(CMakeForceCompiler)
  cmake_force_c_compiler(  "${CROSS_COMPILE}gcc" GNU)
  cmake_force_cxx_compiler("${CROSS_COMPILE}g++" GNU)

endif()



project ("embxx_on_rpi")

set (EXTERNALS_DIR "${CMAKE_SOURCE_DIR}/external")
find_package (Git REQUIRED)

execute_process (
     COMMAND ${GIT_EXECUTABLE} submodule update --init
     WORKING_DIRECTORY ${CMAKE_SOURCE_DIR})

# etc.
```

(personally I'd tend to keep it in `cmake/toolchain_handling.cmake`, but I'd be happy to change it to be all in one `CMakeLists.txt`, if you prefer that.)

> It's also not clear whether enable_language (ASM) needs to be replaced with enable_language (ASM-ATT). You just added a comment there without changing the statement.

`ASM` is the correct setting. If you then check the commands with `make VERBOSE=1`, you'll see that the file `startup.s` is actually compiled with `gcc` (not with `as`!!!). That's cmake default behaviour and desired and all good.
The comment is just there to show that you can still have `startup.s` compiled with `as` (instead of `gcc`), but that that requires `ASM-ATT`.
So ...: if the comment is confusing, we could just remove it.

**Comment by arobenko** [comment]
Hi @ajneu (What' your first name, by the way? How should I address you?)
Yes, I'd like to see the extra toolchain file disappear at this stage, and put everything in the root CMakeLists.txt file. I tend to move things to separate files when there are multiple uses to it to avoid code duplication, otherwise it just hides things that should be visible up front. If you don't mind, please submit a new pull request, or I can use your modifications as a base, and make the required changes myself.

**Comment by ajneu** [comment]
@0xc0170 Here's the truth about the deprecation: there is no deprecation and `CMAKE_FORCE_C_COMPILER` etc are still very much alive for use-cases such as this one (with newlib stubs):
[link](https://github.com/Kitware/CMake/commit/beaa4fa5ddba4298dfc94b38e1695bac28ea5faf) ([also](https://cmake.org/Bug/view.php?id=15797))

**Comment by ajneu** [comment]
@arobenko Hi Alex, I've changed it slightly. I've kept the printing, since `CMAKE_FORCE_CXX_COMPILER` (etc.) does not print the C++ compiler.

Feel free to use this as base.

PS: You can address me as ajneu, or Albert if you prefer.

**Comment by ajneu** [comment]
Hi Alex,

just did another commit.
Brad King just left a very interesting message here: https://cmake.org/Bug/view.php?id=15797#c40536
From CMake 3.5.0 (once released) onwards, one can use 

``` cmake
set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)
```

to allow the following to work, even when written **before** the line with `project ("embxx_on_rpi")`):

``` cmake
set(CMAKE_C_COMPILER   arm-none-eabi-gcc)
set(CMAKE_CXX_COMPILER arm-none-eabi-g++)
```

Previously that brought an error regarding newlib stubs.

I've incorporated into your project as follows:

``` cmake
  set (CMAKE_SYSTEM_NAME "Generic")

  if ("${CROSS_COMPILE}" STREQUAL "")
    set (CROSS_COMPILE "arm-none-eabi-")
  endif()

  if (NOT    CMAKE_VERSION VERSION_LESS 3.5.0)
    set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)
    set(CMAKE_C_COMPILER   "${CROSS_COMPILE}gcc")
    set(CMAKE_CXX_COMPILER "${CROSS_COMPILE}g++")
  else()
    include(CMakeForceCompiler)
    cmake_force_c_compiler(  "${CROSS_COMPILE}gcc" GNU)
    cmake_force_cxx_compiler("${CROSS_COMPILE}g++" GNU)
  endif()
```

To test with 3.5.0 and `CMAKE_TRY_COMPILE_TARGET_TYPE` you need a **very very recent** CMake. More recent than v3.5.0-rc3. Best is to use use it directly [from their repo](https://cmake.org/gitweb?p=cmake.git;a=summary). You can get it like this:

``` bash
## see next and master etc. https://cmake.org/gitweb?p=cmake.git;a=heads

git clone --depth 1 --branch next https://cmake.org/cmake.git cmake_next
cd cmake_next
mkdir build
cd    build
cmake -DCMAKE_INSTALL_PREFIX=$(pwd)/install ..
make -j4
make install
export PATH=$(pwd)/install/bin:$PATH
which cmake
cmake --version  ## should show that very recent cmake
```

**Comment by arobenko** [comment]
Hi Albert,
Let's not rush things. The CMake 3.5 is a long way away until it becomes a mainstream, and even if
it allows better way of doing things, it doesn't mean an old way will immediately stop working.  There is no point to add if-else statement when the "else" part works for both.
My Ubuntu 14.04 LTS is still using CMake 2.8.12.2, and the whole configuration still works for the latest CMake.

 I've merged your previous changes to develop branch with some slight modifications.

**Comment by ajneu** [comment]
Well yes, but setting the compiler (as is possible in 5.3.0 release), is better than forcing it.

**Comment by ajneu** [comment]
Correction: `CMAKE_TRY_COMPILE_TARGET_TYPE` will only be in CMake 3.6 ([ref](https://cmake.org/Bug/view.php?id=15797) see "Target Version"), so it is indeed really quite a long way off still.

**Pull request closed without merging**